src/Symfony/Component/HttpClient/ServerSentEvents/EventSource.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+namespace Symfony\Component\HttpClient\ServerSentEvents;
+
+use Symfony\Component\HttpClient\HttpClient;
+use Symfony\Component\HttpClient\HttpClientTrait;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use Symfony\Contracts\HttpClient\ResponseStreamInterface;
+
+final class EventSource implements HttpClientInterface
+{
+    use HttpClientTrait;
+
+    private $client;
+    private $lastEventId = null;
+    private $reconnectionTime = 5;
+
+    public function __construct(HttpClientInterface $client = null, array $defaultOptions = [])
+    {
+        $this->client = $client ?: HttpClient::create($defaultOptions);
+    }
+
+    public function stream($responses, float $timeout = null): ResponseStreamInterface
+    {
+        return $this->client->stream($responses, $timeout);
+    }
+
+    public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        $options['headers'] = array_merge($options['headers'] ?? [], [
+            'Accept' => 'text/event-stream',
+            'Cache-Control' => 'no-store',
+        ]);
+
+        if ($this->lastEventId) {
+            $options['headers']['Last-Event-ID'] = $this->lastEventId;
+        }
+
+        return $this->client->request($method, $url, $options);
+    }
+
+    public function messages($response): \Iterator
+    {
+        while (true !== $response->getInfo('canceled')) {
+            $buffer = '';
+            foreach ($this->client->stream($response, $this->reconnectionTime) as $chunk) {
+                if ($chunk->isTimeout()) {
+                    yield $chunk;
+                    break;
+                }
+
+                if ($chunk->isFirst()) {
+                    yield $chunk;
+                    continue;
+                }
+
+                // connection closed
+                if ($chunk->isLast() || !$chunk->getContent()) {
+                    yield $chunk;
+                    return null;
+                }
+
+                // We have to gather our own Buffer that may be longer then a chunk, replace byte order mark
+                $buffer .= str_replace("\xEF\xBB\xBF", '', $chunk->getContent());
```

Comment:
Shouldn't the UTF-8 BOM be removed only if it is at the start of the file? What if there is a BOM for another encoding (such as UTF-16)?

Reply:
Also, maybe can you add a comment explaining what this magic sequence is.

Reply:
Actually, according to the spec, the BOM can only be `xFEFF`:

`bom           = %xFEFF ; U+FEFF BYTE ORDER MARK`